### PR TITLE
fix tests and error cases; allow std::optional for streamutils

### PR DESCRIPTION
**Pull request by CharlieL7**
@pfultz2 please check that the change I made to `streamutils` makes sense to allow `optional` to be used for operator attributes. This should make handling optional attributes better.

Fixes a bug I found in allocate and adds testcases.

### CODE REVIEW

File: src/include/migraphx/op/allocate.hpp [hunk 1]
```diff
@@ -62,26 +62,38 @@ struct allocate
 
     shape compute_shape(const std::vector<shape>& inputs) const
     {
-        if(s != shape())
+        if(s)
```
Comment (umangyadav) [inline src/include/migraphx/op/allocate.hpp:65]:
```suggestion
        if(s.has_value())
```
Minor Suggestion : For better readability

File: src/include/migraphx/op/allocate.hpp [hunk 2]
```diff
@@ -62,26 +62,38 @@ struct allocate
...
             else
             {
+                if(s->dynamic())
+                {
+                    MIGRAPHX_THROW("ALLOCATE: dynamic shape attribute and no input");
+                }
                 migraphx::check_shapes{inputs, *this, false}.has(0);
             }
-            return s;
+            return *s;
```
Comment (umangyadav) [inline src/include/migraphx/op/allocate.hpp:83]:
```suggestion
            return s.value();
```

File: src/include/migraphx/op/allocate.hpp [hunk 3]
```diff
@@ -62,26 +62,38 @@ struct allocate
...
+            {
+                MIGRAPHX_THROW("ALLOCATE: shape and buf_type attributes both not set");
+            }
             migraphx::check_shapes{inputs, *this, false}.has(1).only_dims(1);
             const auto& out_dims = inputs.at(0);
             std::size_t max_val = std::numeric_limits<std::size_t>::max();
             std::vector<shape::dynamic_dimension> dyn_dims(out_dims.lens().at(0),
                                                            shape::dynamic_dimension{0, max_val});
-            return {buf_type, dyn_dims};
+            return {*buf_type, dyn_dims};
```
Comment (umangyadav) [inline src/include/migraphx/op/allocate.hpp:96]:
```suggestion
            return {buf_type.value(), dyn_dims};
```

File: src/include/migraphx/op/allocate.hpp [hunk 4]
```diff
@@ -62,26 +62,38 @@ struct allocate
...
+                    MIGRAPHX_THROW("ALLOCATE: dynamic shape attribute and no input");
+                }
                 migraphx::check_shapes{inputs, *this, false}.has(0);
             }
-            return s;
+            return *s;
         }
         else
         {
+            if(not buf_type)
```
Comment (umangyadav) [inline src/include/migraphx/op/allocate.hpp:87]:
```suggestion
            if(not buf_type.has_value())
```

File: src/include/migraphx/op/allocate.hpp [hunk 5]
```diff
@@ -94,7 +106,11 @@ struct allocate
         {
             std::vector<std::size_t> output_dims(output_shape.ndim());
             args.at(0).visit([&](auto a) { output_dims.assign(a.begin(), a.end()); });
-            return argument{shape{buf_type, output_dims}};
+            if(s)
```
Comment (umangyadav) [inline src/include/migraphx/op/allocate.hpp:109]:
```suggestion
            if(s.has_value())
```

File: src/include/migraphx/op/allocate.hpp [hunk 6]
```diff
@@ -62,26 +62,38 @@ struct allocate
...
+            }
             if(inputs.size() == 1)
             {
                 migraphx::check_shapes{inputs, *this, false}.only_dims(1);
             }
             else
             {
+                if(s->dynamic())
+                {
+                    MIGRAPHX_THROW("ALLOCATE: dynamic shape attribute and no input");
```
Comment (umangyadav) [inline src/include/migraphx/op/allocate.hpp:79]: This is inside `else` so technically there could be inputs >=1

File: src/include/migraphx/streamutils.hpp [hunk 7]
```diff
@@ -68,6 +69,13 @@ auto stream_write_value_impl(rank<1>, std::ostream& os, const T& x) -> decltype(
     os << x;
 }
 
+template <class T>
+auto stream_write_value_impl(rank<1>, std::ostream& os, const std::optional<T>& x)
```
Comment (pfultz2) [inline src/include/migraphx/streamutils.hpp:73]: Use `optional<T>` instead of `std::optional<T>`.

File: src/include/migraphx/op/allocate.hpp [hunk 8]
```diff
@@ -48,9 +48,9 @@ namespace op {
  */
 struct allocate
 {
-    shape s{};
+    std::optional<shape> s;
     // for dynamic allocate to set the buffer type
-    shape::type_t buf_type = shape::half_type;
+    std::optional<shape::type_t> buf_type;
```
Comment (pfultz2) [inline src/include/migraphx/op/allocate.hpp:53]: These seems mutually exclusive. Maybe we should use `std::variant` instead , but we will need to update the serialization to support variant.

File: src/include/migraphx/op/allocate.hpp (hunk 1 above)
Comment (bpickrel) [inline src/include/migraphx/op/allocate.hpp:65]: This op. only allows 0 or 1 inputs, correct?  Check that here.

File: src/include/migraphx/op/allocate.hpp [hunk 9]
```diff
@@ -48,9 +48,9 @@ namespace op {
  */
```
Comment (bpickrel) [inline src/include/migraphx/op/allocate.hpp:48]: Suggested modification:

 - * Either `this.s` or `this.buf_type` must be set to calculate the dynamic output shape at compute
 - * time. 

 * For dynamic shapes, either `this.s` or `this.buf_type` must be set to calculate the dynamic output shape at compute
 * time, but not both.

File: src/include/migraphx/op/allocate.hpp
Comment (bpickrel) [inline src/include/migraphx/op/allocate.hpp:1]: Update copyright date

File: src/include/migraphx/op/allocate.hpp (hunk 1 above)
Comment (CharlieL7) [reply src/include/migraphx/op/allocate.hpp:65]: Already checked later on with the check_shapes calls

File: test/ref/allocate.cpp [hunk 10]
```diff
@@ -47,3 +47,21 @@ TEST_CASE(allocate_dyn)
     migraphx::shape sresult{migraphx::shape::float_type, {2, 3, 4, 4}};
     result.visit([&](auto output) { EXPECT(output.get_shape() == sresult); });
 }
+
+TEST_CASE(allocate_dyn1)
+{
+    migraphx::program p;
+    auto* mm = p.get_main_module();
+    migraphx::shape s{migraphx::shape::int64_type, {4}};
+    migraphx::shape out_shape{migraphx::shape::float_type, {2, 3, 4, 4}};
```
[the remark this answers is not in the document]
Comment (CharlieL7) [reply test/ref/allocate.cpp:56]: The `allocate` operator with 1 input outputs a dynamic shape

File: src/include/migraphx/op/allocate.hpp (hunk 6 above)
Comment (CharlieL7) [reply src/include/migraphx/op/allocate.hpp:79]: check_shapes call later limits it to 0 inputs

**Review by pfultz2 — APPROVED** [review]

**Review by umangyadav — APPROVED** [review]

**Pull request merged**